### PR TITLE
Solr8 installer

### PR DESCRIPTION
**Pull request by makuk66**
Add Solr 8.
Changes: http://lucene.apache.org/solr/8_0_0/changes/Changes.html
Being a major release, Solr 8 removes many deprecated APIs, changes various parameter defaults and behavior. Some changes may require a re-index of your content. You are thus encouraged to thoroughly read the "Upgrade Notes". Solr 8.0 also includes many other new features as well as numerous optimizations and bugfixes of the corresponding Apache Lucene release.

On the docker-solr side, switch the on-disk layout for 8 and beyond: instead of the expanded Solr tar, use the `install_solr_service` script. Fixes https://github.com/docker-solr/docker-solr/issues/173

temporary test images:
```
dockersolr/docker-solr:8.0.0-82a2f85
dockersolr/docker-solr:8.0.0-alpine-82a2f85
dockersolr/docker-solr:8.0.0-slim-82a2f85
```

### CODE REVIEW

**Comment by dsmiley** [comment]
I'm all for using the install_solr_service to thus be more standard with upstream.

Why did the changes here touch previous releases?  Wouldn't we want them to stay frozen as they were?  I know with most new releases (all?) you touch old ones too but I don't understand why.

**Comment by risdenk** [comment]
FWIW did some quick Solr S3 testing here based on this image:
https://github.com/risdenk/solr-s3a-testing

**Comment by makuk66** [comment]
> Why did the changes here touch previous releases? Wouldn't we want them to stay frozen as they were? I know with most new releases (all?) you touch old ones too but I don't understand why.

For practicality really. When I update things in Dockerfiles or scripts, they are typically fixes or improvements, and I'd like those to be available to all 16 versions that the Docker Library team is building. To ensure that happens consistently, I use the `tools/update.sh` script which renders the Dockerfile templates and copies the scripts to all version and variant directories. If I didn't have that, I'd have to manually make those changes, and would be more likely to make mistakes. This has been working really well, when we've had one consistent way in which the scripts work. Now that there is a different way for 8, versus prior versions, it's a little more complicated, but by expressing the differences in a checked-in script, it is easier to understand that they are intentional, and they are easier to manage, than if I had to remember manual changes. We could implement that differently, say in this case have separate `scripts/old-style` and `scripts/service-installer` directories, and the same for the test directories, but then you end up with duplicated files, which risk becoming out of sync or out of date, and that becomes unwieldy.

We could alternatively throw out all older versions, so they do not get rebuilt (but remain on Docker hub), and we don't have to worry about maintaining them; but that would mean users of those versions would not receive security fixes and other improvements in the base images.

Basically, if you want to absolutely stick to un-changing bits-in-the-image for a specific installation, for the ultimate in repeatability, you should tag a specific image, and then host that yourself in a private registry. Then you can update/test/deploy that explicitly, at a schedule that suits you.

But yeah, it always looks scary when there are 162 files changed. When reviewing, you're best off to look changes in the templates, the `scripts/` and the `tools/`, and then just scan over the copies in the versions/variants.

**Comment by dsmiley** [comment]
The main risk I see is testing to ensure we don't introduce a regression.  How confident are we that our generic scripts don't break some peculiarity of 5.x?  Especially with minutia like the ability to use an environment variable to override a default value.  I don't know.  Have we screwed up before?

**Comment by makuk66** [comment]
> The main risk I see is testing to ensure we don't introduce a regression. How confident are we that our generic scripts don't break some peculiarity of 5.x? Especially with minutia like the ability to use an environment variable to override a default value. I don't know. Have we screwed up before?

When change happens you can't rule out regressions (and Solr has its own too); which is why I say that if zero regression is crucial to you, you should tag and host images that you have vetted, and that applies to all docker images. Like any software project, if you do make changes, a regression can happen, and then you have to deal with it, either by changing the software, or how it is used. We have some minimal testing to spot major issues, which should help some.

The question is, what do you do. I have explored the alternative approach I mentioned, see https://github.com/docker-solr/docker-solr/compare/solr-8-installer-copied2?expand=1 . There we have no changes to the pre-8 files, at the cost of duplication. I think that's fine too.

**Comment by dsmiley** [comment]
Personally I prefer #209 since there is a notable change in this major release in where we put the files, and I don't think we should disturb the previous releases.  That said, I sympathize with your arguments and won't argue it any further.  I wonder what the "normal" / typical approach to this is on other projects?  I dunno; just thinking aloud.

Any way, thanks for doing all this work!

**Comment by risdenk** [comment]
@makuk66 curious what the status is here? Was surprised to see that 8.0.0 wasn't released yet. Looks like might have been held up by 6.6.6 release?

Let me know if there is anything I can help with.

**Comment by makuk66** [comment]
During further testing I discovered that Solr's `install_solr_service.sh` causes us problems:

- it does not support Alpine
- it does init stuff that is not relevant to containers
- it ignores errors

The [OS detection](https://github.com/apache/lucene-solr/blob/master/solr/bin/install_solr_service.sh#L62) tries to determine the OS in various ways until it sees something it recognises. When running on an unknown distro under Docker, it ends up looking at the running kernel (which is the host's), and that may confuse it. For example, when running an Alpine container on an Ubuntu host, it will spot the "Ubuntu" compiler in the kernel version line, and conclude wrongly it is running on Ubuntu. Whereas when running on Docker on Mac, an Alpine container runs on a Alpine VM, the kernel line only contains Alpine, and you get an "Unable to auto-detect" error.

The [auto-start service](https://github.com/apache/lucene-solr/blob/master/solr/bin/install_solr_service.sh#L354) is not something we want in a container. The command fails, because the container do not have that command. But the script ignores errors, and just carries on, so it kinda works, but it's a bit of a fudge.

The script also allows for various overrides, and different dist files and so on, which are not needed for our use-case.

So I am now not all that keen on using it. I'm bummed about that, because I figured this was done, and it has cost me enough time already. The alternative is to doing an install the same way but without the actual script, and then we can simply put it in the Dockerfile for clarity and see what is going on. Working through that took a while, but seems to be in a reasonable place now, see https://github.com/docker-solr/docker-solr/pull/210

**Comment by risdenk** [comment]
@makuk66 ah that is frustrating. I didn't test the Alpine integration. Thanks for the details.
